File: src/core/execution/nodes/telegram_memory_selector.py
"""
Telegram Memory Selector Node
Retrieves relevant context for a Telegram chat
"""
import time
from typing import Dict, Any, Optional, List
from ..node_base import BaseNode, ExecutionContext
from ....utils.logger import get_logger

logger = get_logger(__name__)


class TelegramMemorySelectorNode(BaseNode):
    """
    Retrieves relevant context for a Telegram chat.
    
    Fetches:
    - Recent messages from the specific chat
    - Summaries created for this chat
    - Combines them into usable context for the agent
    
    Inputs:
        message: The incoming message to find context for (optional, for semantic search)
        chat_id: Telegram chat/group ID to filter by (required)
        storage_instance: StorageInterface instance (required)
        model: ObeliskLLM instance (optional, for semantic search)
    
    Properties:
        recent_count: Number of recent messages to include (default: 20)
        include_summaries: Whether to include summaries (default: true)
    
    Outputs:
        context: Combined context (summaries + recent messages)
        recent_messages: Just the recent raw messages
        summaries: Just the summaries
        message: Original message passed through (for chaining to next node)
    """
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        """Initialize telegram memory selector node"""
        super().__init__(node_id, node_data)
        self._recent_count = int(self.metadata.get('recent_count', 20))
        self._include_summaries = self.metadata.get('include_summaries', True)
    
    def _get_recent_messages(
        self,
        storage_instance,
        chat_id: str,
        count: int
    ) -> List[Dict]:
        """Fetch recent messages for a chat from storage"""
        try:
            # Get activity logs filtered by type and chat_id
            logs = storage_instance.get_activity_logs(
                activity_type='telegram_message',
                limit=count * 2  # Get more than needed, we'll filter by chat_id
            )
            
            # Filter by chat_id and take most recent
            chat_messages = []
            for log in logs:
                metadata = log.get('metadata', {})
                if metadata.get('chat_id') == chat_id:
                    chat_messages.append({
                        'message': metadata.get('message', ''),
                        'user_id': metadata.get('user_id', ''),
                        'username': metadata.get('username', ''),
                        'timestamp': metadata.get('timestamp', log.get('timestamp', 0))
                    })
                    if len(chat_messages) >= count:
                        break
            
            # Sort by timestamp (newest last for chronological order)
            chat_messages.sort(key=lambda x: x.get('timestamp', 0))
            
            return chat_messages
            
        except Exception as e:
            logger.error(f"[TelegramMemorySelector] Error fetching messages: {e}")
            return []
    
    def _get_summaries(
        self,
        storage_instance,
        chat_id: str,
        limit: int = 5
    ) -> List[Dict]:
        """Fetch summaries for a chat from storage"""
        try:
            # Get activity logs for summaries
            logs = storage_instance.get_activity_logs(
                activity_type='telegram_summary',
                limit=limit * 2
            )
            
            # Filter by chat_id
            chat_summaries = []
            for log in logs:
                metadata = log.get('metadata', {})
                if metadata.get('chat_id') == chat_id:
                    chat_summaries.append({
                        'summary': metadata.get('summary', ''),
                        'keyTopics': metadata.get('keyTopics', []),
                        'sentiment': metadata.get('sentiment', ''),
                        'timestamp': metadata.get('timestamp', log.get('timestamp', 0))
                    })
                    if len(chat_summaries) >= limit:
                        break
            
            # Sort by timestamp (most recent first)
            chat_summaries.sort(key=lambda x: x.get('timestamp', 0), reverse=True)
            
            return chat_summaries
            
        except Exception as e:
            logger.error(f"[TelegramMemorySelector] Error fetching summaries: {e}")
            return []
    
    def _format_messages(self, messages: List[Dict]) -> str:
        """Format messages into readable text"""
        if not messages:
            return ""
        
        lines = ["=== Recent Messages ==="]
        for msg in messages:
            username = msg.get('username') or msg.get('user_id', 'Unknown')
            text = msg.get('message', '')
            lines.append(f"[{username}]: {text}")
        
        return "\n".join(lines)
    
    def _format_summaries(self, summaries: List[Dict]) -> str:
        """Format summaries into readable text"""
        if not summaries:
            return ""
        
        lines = ["=== Chat Summaries ==="]
        for i, summary in enumerate(summaries, 1):
            lines.append(f"\n--- Summary {i} ---")
            lines.append(summary.get('summary', 'No summary available'))
            topics = summary.get('keyTopics', [])
            if topics:
                lines.append(f"Topics: {', '.join(topics)}")
            sentiment = summary.get('sentiment', '')
            if sentiment:
                lines.append(f"Sentiment: {sentiment}")
        
        return "\n".join(lines)
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """Execute telegram memory selector - retrieves context for a chat"""
        message = self.get_input_value('message', context, '')  # Optional
        chat_id = self.get_input_value('chat_id', context, '')
        storage_instance = self.get_input_value('storage_instance', context, None)
        llm = self.get_input_value('model', context, None)  # Optional for semantic search
        
        # Get properties
        recent_count = int(self.metadata.get('recent_count', 20))
        include_summaries = self.metadata.get('include_summaries', True)
        
        # Validate required inputs
        if not chat_id:
            logger.warning("[TelegramMemorySelector] No chat_id provided")
            return {'context': '', 'recent_messages': '', 'summaries': '', 'message': str(message) if message else ''}
        
        if storage_instance is None:
            raise ValueError("storage_instance is required for TelegramMemorySelectorNode")
        
        # Fetch recent messages
        messages = self._get_recent_messages(storage_instance, str(chat_id), recent_count)
        recent_messages_text = self._format_messages(messages)
        
        # Fetch summaries if enabled
        summaries_text = ""
        if include_summaries:
            summaries = self._get_summaries(storage_instance, str(chat_id))
            summaries_text = self._format_summaries(summaries)
        
        # Combine into context
        context_parts = []
        if summaries_text:
            context_parts.append(summaries_text)
        if recent_messages_text:
            context_parts.append(recent_messages_text)
        
        combined_context = "\n\n".join(context_parts) if context_parts else "No chat history available."
        
        logger.info(f"[TelegramMemorySelector] Retrieved {len(messages)} messages and {len(summaries) if include_summaries else 0} summaries for chat {chat_id}")
        if messages:
            logger.debug(f"[TelegramMemorySelector] Sample message: {messages[-1]}")
        logger.debug(f"[TelegramMemorySelector] Context length: {len(combined_context)} chars")
        
        return {
            'context': combined_context,
            'recent_messages': recent_messages_text,
            'summaries': summaries_text,
            'message': str(message) if message else ''
        }

File: ts/src/core/execution/nodes/telegramMemorySelector.ts
/**
 * TelegramMemorySelectorNode – retrieves relevant context for a Telegram chat.
 * Mirrors Python src/core/execution/nodes/telegram_memory_selector.py
 *
 * When multiple summaries are available, uses the LLM (non-thinking) to select
 * the most relevant ones — just like the regular MemorySelectorNode.
 *
 * Inputs:
 *   message: The incoming message (optional, passed through)
 *   chat_id: Telegram chat/group ID to filter by (required)
 *   storage_instance: StorageInterface instance (required)
 *   model: InferenceClient (optional, used for intelligent summary selection)
 *
 * Properties:
 *   recent_count: Number of recent messages to include (default: 20)
 *   include_summaries: Whether to include summaries (default: true)
 *
 * Outputs:
 *   context: Combined context string (selected summaries + recent messages)
 *   recent_messages: Formatted recent messages string
 *   summaries: Formatted summaries string
 *   message: Original message passed through (for chaining to next node)
 */
import { BaseNode, ExecutionContext } from "../nodeBase";
import { StorageInterface } from "../../types";
import { InferenceClient } from "./inference/inferenceClient";
import { extractJsonFromLlmResponse } from "../../../utils/jsonParser";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("telegramMemorySelector");

interface ChatMessage {
  message: string;
  user_id: string;
  username: string;
  timestamp: number;
}

interface ChatSummary {
  summary: string;
  keyTopics: string[];
  sentiment: string;
  timestamp: number;
  importantMessages?: string[];
  activeUsers?: string[];
}

export class TelegramMemorySelectorNode extends BaseNode {
  private _recentCount: number;
  private _includeSummaries: boolean;

  constructor(nodeId: string, nodeData: import("../../types").NodeData) {
    super(nodeId, nodeData);
    this._recentCount = Number(this.metadata.recent_count ?? 20);
    this._includeSummaries = this.metadata.include_summaries !== false;
  }

  async execute(context: ExecutionContext): Promise<Record<string, unknown>> {
    const message = this.getInputValue("message", context, "") as string;
    const chatId = this.getInputValue("chat_id", context, "") as string;
    const storageInstance = this.getInputValue(
      "storage_instance",
      context,
      undefined
    ) as StorageInterface | undefined;
    const model = this.getInputValue("model", context, undefined) as
      | InferenceClient
      | undefined;

    // Validate required inputs
    if (!chatId) {
      logger.warning("[TelegramMemorySelector] No chat_id provided");
      return {
        context: "",
        recent_messages: "",
        summaries: "",
        message: message ? String(message) : "",
      };
    }

    if (!storageInstance) {
      throw new Error(
        "storage_instance is required for TelegramMemorySelectorNode"
      );
    }

    // Fetch recent messages
    const messages = await this._getRecentMessages(
      storageInstance,
      String(chatId),
      this._recentCount
    );
    const recentMessagesText = this._formatMessages(messages);

    // Fetch summaries if enabled — with LLM-based selection when available
    let summariesText = "";
    let summaryCount = 0;
    if (this._includeSummaries) {
      const allSummaries = await this._getSummaries(
        storageInstance,
        String(chatId),
        10 // Fetch more than we need so we can select
      );
      summaryCount = allSummaries.length;

      if (allSummaries.length > 0) {
        let selectedSummaries: ChatSummary[];

        // Use LLM to select relevant summaries when we have more than topK and a model
        const topK = 5;
        if (allSummaries.length > topK && model && message) {
          logger.info(
            `[TelegramMemorySelector] Using LLM to select ${topK} relevant summaries from ${allSummaries.length} candidates`
          );
          selectedSummaries = await this._selectRelevantSummaries(
            model,
            String(message),
            allSummaries,
            topK
          );
        } else {
          if (allSummaries.length > 1) {
            logger.info(
              `[TelegramMemorySelector] ${allSummaries.length} summaries (≤${topK}) — using all without LLM selection`
            );
          }
          selectedSummaries = allSummaries;
        }

        summariesText = this._formatSummaries(selectedSummaries);
      }
    }

    // Combine into context
    const contextParts: string[] = [];
    if (summariesText) contextParts.push(summariesText);
    if (recentMessagesText) contextParts.push(recentMessagesText);

    const combinedContext = contextParts.length
      ? contextParts.join("\n\n")
      : "No chat history available.";

    logger.info(
      `[TelegramMemorySelector] Retrieved ${messages.length} messages and ${summaryCount} summaries for chat ${chatId}`
    );

    return {
      context: combinedContext,
      recent_messages: recentMessagesText,
      summaries: summariesText,
      message: message ? String(message) : "",
    };
  }

  // ── Private helpers ────────────────────────────────────────────────

  /**
   * Use LLM (non-thinking) to select the most relevant summaries for the
   * current message. Mirrors the regular MemorySelectorNode._selectRelevantMemories.
   */
  private async _selectRelevantSummaries(
    model: InferenceClient,
    userMessage: string,
    summaries: ChatSummary[],
    topK: number
  ): Promise<ChatSummary[]> {
    if (!summaries.length) return [];
    if (summaries.length <= topK) return summaries;

    try {
      // Format summaries for LLM analysis
      let summariesText = "";
      for (let i = 0; i < summaries.length; i++) {
        const s = summaries[i];
        let str = `Summary ${i}:\n`;
        str += `  Content: ${s.summary || "N/A"}\n`;
        if (s.keyTopics.length) {
          str += `  Topics: ${s.keyTopics.join(", ")}\n`;
        }
        if (s.sentiment) {
          str += `  Sentiment: ${s.sentiment}\n`;
        }
        if (s.importantMessages?.length) {
          str += `  Key messages: ${s.importantMessages.slice(0, 3).join("; ")}\n`;
        }
        summariesText += str + "\n";
      }

      const systemPrompt = `You are a memory selector for a Telegram chat agent. Your role is to select the ${topK} most relevant chat summaries given the current incoming message.

You MUST return ONLY valid JSON. No markdown code blocks, no explanations, no text before or after the JSON. Start with { and end with }.

Analyze which summaries are most relevant and return a JSON object with:
- selected_indices: Array of 0-based indices of the ${topK} most relevant summaries (e.g., [0, 2, 5])
- reason: Brief explanation of why these summaries were selected

Example of correct JSON format:
{
  "selected_indices": [0, 2, 5],
  "reason": "Summary 0 discusses the main topic, Summary 2 has relevant context"
}`;

      const query = `Current message: ${userMessage}\n\nAvailable Chat Summaries:\n${summariesText}\nReturn the indices (0-based) of the ${topK} most relevant summaries. Return ONLY the JSON object, nothing else.`;

      logger.info(
        `[TelegramMemorySelector] Calling inference for summary selection (${summaries.length} candidates)`
      );

      const result = await model.generate(
        query,
        systemPrompt,
        0.1, // Low quantum_influence for consistent selection
        800,
        null,
        false // Non-thinking for fast selection
      );

      const selectionText = (result.response ?? "").trim();
      const selectionData = extractJsonFromLlmResponse(
        selectionText,
        "telegram_memory_selection"
      );

      const selectedIndices =
        (selectionData.selected_indices as number[]) ?? [];

      const selectedSummaries: ChatSummary[] = [];
      for (const idx of selectedIndices) {
        if (typeof idx === "number" && idx >= 0 && idx < summaries.length) {
          selectedSummaries.push(summaries[idx]);
        }
      }

      if (selectedSummaries.length) {
        logger.info(
          `[TelegramMemorySelector] LLM selected ${selectedSummaries.length} relevant summaries from ${summaries.length} total`
        );
        return selectedSummaries;
      }

      // Fallback: return first topK if LLM selection returned invalid indices
      logger.warning(
        `[TelegramMemorySelector] LLM selection returned invalid indices: ${JSON.stringify(selectedIndices)}, falling back to first ${topK}`
      );
      return summaries.slice(0, topK);
    } catch (err) {
      logger.error(
        `[TelegramMemorySelector] Error in LLM summary selection: ${err}`
      );
      // Fallback: return first topK
      return summaries.slice(0, topK);
    }
  }

  private async _getRecentMessages(
    storage: StorageInterface,
    chatId: string,
    count: number
  ): Promise<ChatMessage[]> {
    try {
      const logs = await storage.getActivityLogs("telegram_message", count * 2);
      const chatMessages: ChatMessage[] = [];

      for (const log of logs) {
        const meta = (log.metadata ?? {}) as Record<string, unknown>;
        if (String(meta.chat_id ?? "") === chatId) {
          chatMessages.push({
            message: String(meta.message ?? ""),
            user_id: String(meta.user_id ?? ""),
            username: String(meta.username ?? ""),
            timestamp: Number(
              meta.timestamp ?? (log.created_at ? new Date(log.created_at).getTime() / 1000 : 0)
            ),
          });
          if (chatMessages.length >= count) break;
        }
      }

      // Sort by timestamp (newest last for chronological order)
      chatMessages.sort((a, b) => a.timestamp - b.timestamp);
      return chatMessages;
    } catch (err) {
      logger.error(
        `[TelegramMemorySelector] Error fetching messages: ${err}`
      );
      return [];
    }
  }

  private async _getSummaries(
    storage: StorageInterface,
    chatId: string,
    limit = 10
  ): Promise<ChatSummary[]> {
    try {
      const logs = await storage.getActivityLogs(
        "telegram_summary",
        limit * 2
      );
      const chatSummaries: ChatSummary[] = [];

      for (const log of logs) {
        const meta = (log.metadata ?? {}) as Record<string, unknown>;
        if (String(meta.chat_id ?? "") === chatId) {
          chatSummaries.push({
            summary: String(meta.summary ?? ""),
            keyTopics: (meta.keyTopics as string[]) ?? [],
            sentiment: String(meta.sentiment ?? ""),
            importantMessages: (meta.importantMessages as string[]) ?? [],
            activeUsers: (meta.activeUsers as string[]) ?? [],
            timestamp: Number(
              meta.timestamp ?? (log.created_at ? new Date(log.created_at).getTime() / 1000 : 0)
            ),
          });
          if (chatSummaries.length >= limit) break;
        }
      }

      // Sort by timestamp (most recent first)
      chatSummaries.sort((a, b) => b.timestamp - a.timestamp);
      return chatSummaries;
    } catch (err) {
      logger.error(
        `[TelegramMemorySelector] Error fetching summaries: ${err}`
      );
      return [];
    }
  }

  private _formatMessages(messages: ChatMessage[]): string {
    if (!messages.length) return "";
    const lines = ["=== Recent Messages ==="];
    for (const msg of messages) {
      const username = msg.username || msg.user_id || "Unknown";
      lines.push(`[${username}]: ${msg.message}`);
    }
    return lines.join("\n");
  }

  private _formatSummaries(summaries: ChatSummary[]): string {
    if (!summaries.length) return "";
    const lines = ["=== Chat Summaries ==="];
    for (let i = 0; i < summaries.length; i++) {
      const s = summaries[i];
      lines.push(`\n--- Summary ${i + 1} ---`);
      lines.push(s.summary || "No summary available");
      if (s.keyTopics.length) {
        lines.push(`Topics: ${s.keyTopics.join(", ")}`);
      }
      if (s.sentiment) {
        lines.push(`Sentiment: ${s.sentiment}`);
      }
    }
    return lines.join("\n");
  }
}
